move kwarg parsing to api backend instead of in pepper

mattp- opened the pull request:
rather than attempting to deconstruct args locally to fully reproduce
the proper low state to pass to the api, we instead leave the kwargs as
is. The reason for this is the RunnerClient will deconstruct and
deserialize nested yaml args. We could equivalently do that locally but
it seems pepper tries to be a no-dependency dist so we just hand off the
work instead.

This PR partially depends on release of https://github.com/saltstack/salt/pull/50124
to determine if a release is neon or newer; however, we can safely assume that if
the salt_version header is not provided in a request it is a salt-api
older than that.

Fixes #57

mattp- committed:
move kwarg parsing to api backend instead of in pepper

rather than attempting to deconstruct args locally to fully reproduce
the proper low state to pass to the api, we instead leave the kwargs as
is. The reason for this is the RunnerClient will deconstruct and
deserialize nested yaml args. We could equivalently do that locally but
it seems pepper tries to be a no-dependency dist so we just hand off the
work instead.

This PR partially depends on release of https://github.com/saltstack/salt/pull/50124
to determine if a release is neon or newer; however, we can safely assume that if
the salt_version header is not provided in a request it is a salt-api
older than that.

Fixes #57

gtmanfred commented:
@mattp- tests /should/ be fixed now

gtmanfred commented on pepper/libpepper.py:
```diff
@@ -231,6 +233,10 @@ def req(self, path, data=None):
             ret = json.loads(content)
+
+            if not self.salt_version and 'x-salt-version' in f.headers:
+                self._parse_salt_version(f.headers['x-salt-version'])
```
Can you write some mocked unittests that test this feature out?

mattp- committed:
we should be testing tornado as well as cherrypy

mattp- committed:
test tornado and cherrypy

I originally tried making this work as a pytest parameterized fixture,
but couldn't get things working properly with pytest-salt daemon mgmt.
If someone figures that out feel free to revert this tox level knob.

mattp- committed:
make verbose setting map to root logger

was being applied pepper.cli before which is useless for
pepper.libpepper etc.

mattp- committed:
use https instead of git:// in requirements

mattp- committed:
tox.ini add missing indentation

mattp- committed:
fix broken wheel test

minions.connected doesnt take an arg

mattp- committed:
fix broken poller tests

failed output isnt returned if nothing failed anymore. additionally,
returned as valid json

mattp- committed:
improve /run handling, token handling with /run

make this act more reasonable. we drop the x-auth-token header when
using /run as the idea of that endpoint is to use whats in the lowstate.
we get rid of the /token route handling as it seemingly does the same
thing as /login, and doesnt actually exist in the rest_tornado impl so
I'm considering it deprecated for simplicity.

mattp- committed:
Improve ret handling across cherrypy/tornado

this seems to do the right thing in more scenarios.

mattp- committed:
add showlocals to tox.ini

mattp- committed:
Revert "improve /run handling, token handling with /run"

This reverts commit 59c013ab792f1abf3a76d1d0a224db7bbe1e8864.

this fixes tornado but breaks cherrypy - I mistakenly assumed the
payload token returned on login was the real eauth token not the
session.id. I'll add a route to /token in tornado that is equivalent to
/login on tornado to mimick backcompat behavior even though they are
equivalent there due to not having a proxy session store.

mattp- committed:
don't pass an x-auth-header when /run

auth should be handled only by token provided in low

mattp- committed:
make outputter behave more like salt

runner output should be identical now

mattp- committed:
parameterize and improve /run | /login handling

both codepaths are valid in both cherrypy and tornado. bunch of stuff
xfail'd for now in tornado as there are pending develop/ PR's to salt to
make things fully cross-compatible.

mattp- committed:
pass along timeout kwarg in low

this will be used by tornado to know how long to persist sync calls for
instead of the app_default_timeout. Also seems like a reasonable thing
to do anyway.

mattp- committed:
raise logged exception failures as errors

mattp- committed:
this needs xfail on tornado as well

mattp- committed:
let tox test multiple salt versions, both cherrypy and tornado backends

mattp- committed:
Merge remote-tracking branch 'upstream/develop' into develop
